# Applied Machine Learning — Lab Assignment 1

'''**Course:** CSAI2017P — Applied Machine Learning (Lab)  
**Name:** Abhishek Bhatt  
**SAP ID:** 590028847  
**Batch:** B-05

**Experiments:** 1–2  
**Topic:** Environment Setup, Preprocessing and Exploratory Data Analysis
'''

In [1]:
import sys
import numpy as np
import pandas as pd
import matplotlib
import matplotlib.pyplot as plt
import seaborn as sns
import sklearn

from sklearn.datasets import fetch_california_housing
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsRegressor
from sklearn.metrics import mean_absolute_error
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler

print("Python:", sys.version.split()[0])
print("NumPy:", np.__version__)
print("Pandas:", pd.__version__)
print("Scikit-learn:", sklearn.__version__)
print("Matplotlib:", matplotlib.__version__)
print("Seaborn:", sns.__version__)

Python: 3.14.2
NumPy: 2.5.2
Pandas: 3.0.5
Scikit-learn: 1.9.0
Matplotlib: 3.11.1
Seaborn: 0.13.2


### Observation (A1)

All required libraries were installed and imported successfully. Version pinning
helps ensure that the same code behaves consistently when the work is run on another system.

## A2 — First Look at the Data

**Dataset:** California Housing

In [2]:
X, y = fetch_california_housing(return_X_y=True, as_frame=True)

df = X.copy()
df["MedHouseVal"] = y

print("Rows:", df.shape[0])
print("Features:", X.shape[1])
print("Target:", y.name)

Rows: 20640
Features: 8
Target: MedHouseVal


In [3]:
print("First 5 rows:")
display(df.head())

print("\nDataset information:")
df.info()

print("\nDescriptive statistics:")
display(df.describe())

print("\nMissing values:")
display(df.isna().sum())

First 5 rows:


,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude,MedHouseVal
0,8.3252,41.0,6.984127,1.023810,322.0,2.555556,37.88,-122.23,4.526
1,8.3014,21.0,6.238137,0.971880,2401.0,2.109842,37.86,-122.22,3.585
2,7.2574,52.0,8.288136,1.073446,496.0,2.802260,37.85,-122.24,3.521
3,5.6431,52.0,5.817352,1.073059,558.0,2.547945,37.85,-122.25,3.413
4,3.8462,52.0,6.281853,1.081081,565.0,2.181467,37.85,-122.25,3.422



Dataset information:
<class 'pandas.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   MedInc       20640 non-null  float64
 1   HouseAge     20640 non-null  float64
 2   AveRooms     20640 non-null  float64
 3   AveBedrms    20640 non-null  float64
 4   Population   20640 non-null  float64
 5   AveOccup     20640 non-null  float64
 6   Latitude     20640 non-null  float64
 7   Longitude    20640 non-null  float64
 8   MedHouseVal  20640 non-null  float64
dtypes: float64(9)
memory usage: 1.4 MB

Descriptive statistics:


,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude,MedHouseVal
count,20640.000000,20640.000000,20640.000000,20640.000000,20640.000000,20640.000000,20640.000000,20640.000000,20640.000000
mean,3.870671,28.639486,5.429000,1.096675,1425.476744,3.070655,35.631861,-119.569704,2.068558
std,1.899822,12.585558,2.474173,0.473911,1132.462122,10.386050,2.135952,2.003532,1.153956
min,0.499900,1.000000,0.846154,0.333333,3.000000,0.692308,32.540000,-124.350000,0.149990
25%,2.563400,18.000000,4.440716,1.006079,787.000000,2.429741,33.930000,-121.800000,1.196000
50%,3.534800,29.000000,5.229129,1.048780,1166.000000,2.818116,34.260000,-118.490000,1.797000
75%,4.743250,37.000000,6.052381,1.099526,1725.000000,3.282261,37.710000,-118.010000,2.647250
max,15.000100,52.000000,141.909091,34.066667,35682.000000,1243.333333,41.950000,-114.310000,5.000010



Missing values:


MedInc         0
HouseAge       0
AveRooms       0
AveBedrms      0
Population     0
AveOccup       0
Latitude       0
Longitude      0
MedHouseVal    0
dtype: int64

### Observation (A2)

The dataset contains 20,640 rows and 8 features. The target is `MedHouseVal`,
measured in $100,000 units. `Population` is on a much larger scale, with a
mean of 1425.48 compared with 28.64 for `HouseAge` and 5.43 for `AveRooms`.
No missing values were found.

## B1 — Scaling Changes the Answer

**Dataset:** California Housing

In [4]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=0
)

knn_raw = KNeighborsRegressor(n_neighbors=5)
knn_raw.fit(X_train, y_train)

y_pred_raw = knn_raw.predict(X_test)
mae_raw = mean_absolute_error(y_test, y_pred_raw)

print(f"MAE without scaling: {mae_raw:.4f} ($100,000 units)")
print(f"Approximately ${mae_raw * 100000:,.0f}")

MAE without scaling: 0.8142 ($100,000 units)
Approximately $81,416


In [5]:
knn_scaled = Pipeline([
    ("scaler", StandardScaler()),
    ("knn", KNeighborsRegressor(n_neighbors=5))
])

knn_scaled.fit(X_train, y_train)

y_pred_scaled = knn_scaled.predict(X_test)
mae_scaled = mean_absolute_error(y_test, y_pred_scaled)

print(f"MAE with scaling: {mae_scaled:.4f} ($100,000 units)")
print(f"Approximately ${mae_scaled * 100000:,.0f}")

MAE with scaling: 0.4308 ($100,000 units)
Approximately $43,076


### Observation (B1)

The MAE decreased from 0.8142 to 0.4308 after scaling. KNN uses distances,
so features with larger numerical scales can have greater influence on the
distance calculation. StandardScaler puts the features on a comparable scale,
which changes the nearest neighbours and improves the predictions.

## C1 — Reproducibility Check

In [6]:
print("Random state used for the train/test split: 0")
print(f"MAE without scaling: {mae_raw:.4f}")
print(f"MAE with scaling: {mae_scaled:.4f}")

Random state used for the train/test split: 0
MAE without scaling: 0.8142
MAE with scaling: 0.4308


### Observation (C1)

After restarting the kernel and running the notebook from top to bottom,
the reported values were reproduced exactly. The fixed `random_state=0`
ensures that the same train/test split is used each time.